### PR TITLE
[dask-on-ray] Add multiple return DataFrame shuffle optimization.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Adds support for multiple return tasks in the Dask-on-Ray scheduler, and adds a DataFrame shuffle optimization that uses multiple return tasks.

## Why are these changes needed?

Changes are needed to speed up large-scale DataFrame shuffling.

## Related issue number

https://github.com/ray-project/ray/issues/13959

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/dask/optimizations.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        if istask(v) and v[0] == shuffle_group:
+            task_deps = dependencies[k]
+            # Only rewrite shuffle group split if all downstream dependencies
+            # are splits.
+            if all(
+                    istask(dsk[dep]) and dsk[dep][0] == operator.getitem
+                    for dep in task_deps):
+                for dep in task_deps:
+                    # Rewrite split
+                    pass
```
Comment (clarkzinzow) [inline python/ray/util/dask/optimizations.py:168]: I'm planning on making this more generic implementation work in the near future, so I'm slightly tempted to keep this here, despite it not doing anything.

**Comment by rkooo567** [comment]
Will review all of them by the end of this weekends! Thanks again for this @clarkzinzow Really cool feature :)

File: python/ray/util/dask/optimizations.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/util/dask/optimizations.py:168]: Is generic implementation something we discussed before, or are you thinking something else?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
tests:test_dask_scheduler => This fails. Also, can you add an unit test that runs some shuffle workload? (e.g., set_index)

**Comment by clarkzinzow** [comment]
@rkooo567 I fixed that failing test in [this PR](https://github.com/ray-project/ray/pull/13945), which has been merged into master. I'll rebase onto master and get that test passing in this branch, but I don't think that it should be seen as a blocker for merging this.

File: python/ray/tests/test_dask_optimization.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            npartitions=npartitions)
+        # We set max_branch=npartitions in order to ensure that the task-based
+        # shuffle happens in a single stage, which is required in order for our
+        # optimization to work.
+        a = df.set_index(
+            ["age"], shuffle="tasks", max_branch=npartitions).head(
+                10, npartitions=-1, compute=False)
+        a.compute()
+
+    assert mock_rewrite.call_count == 2
```
Comment (ericl) [inline python/ray/tests/test_dask_optimization.py:58]: Nice!

Comment (clarkzinzow) [reply python/ray/tests/test_dask_optimization.py:58]: I was originally hoping to also test
```
mock_rewrite.assert_called_with(a.__dask_graph__(), keys=a.__dask_keys__())
```
but the graph is needlessly copied [here](https://github.com/dask/dask/blob/2640241fbdf0c5efbcf35d96eb8cc9c3df4de2fd/dask/base.py#L342) and equality is judged by reference. 😞

File: python/ray/util/dask/optimizations.py (hunk 1 above)
Comment (clarkzinzow) [reply python/ray/util/dask/optimizations.py:160]: It's pretty much what we discussed before!

File: python/ray/tests/test_dask_optimization.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
```
Comment (rkooo567) [inline python/ray/tests/test_dask_optimization.py:57]: Can you also write a simple assert to make sure the df is sorted?

**Review by rkooo567 — COMMENTED** [review]
LGTM. One last comment.

**Review by rkooo567 — APPROVED** [review]

**Comment by ericl** [comment]
Lint and tests

**Comment by clarkzinzow** [comment]
Failing tests are unrelated to this PR, I think this is probably good to go.

**Pull request merged**